paint/PAINT_DrawWindow.cpp

```diff
@@ -55,13 +59,20 @@ bool DrawWindow::mouseButtonUp(win::MouseButton const b)
 {
 	if (b == MouseButton::Left) {
 		if (activeTool_) {
-			activeTool_->clearLines();
+			activeTool_->toolFunctionEnd(mouseCoords_, prevMouseCoords_, startCoord_, this->getRect());
 		}
 	}
 
 	return false;
 }
 
+/*override*/
+bool DrawWindow::mouseExit(MouseButton button)
+{
+	mouseButtonUp(button);
```

Comment:
Calling a virtual function is more expensive than calling a non-virtual member function. That's because under the hood they are stored as pointers in a table (called a 'vtable') that each instance of the class carries around with it. Every time you call a virtual function, the appropriate pointer has to be looked up in the table, and there is a small but sometimes significant time penalty for doing that.

Solution: make both the handlers call a (preferably static) function

```
static void handleMouseUp(MouseButton b, Tool * tool, Coord const & mouse, Coord const & prevMouse, Coord const & start, Rect const & rect)
{
    if (b == MouseButton::Left && tool) {
        tool->toolFunctionEnd(mouse, prevMouse, start, rect);
   }
}

void DrawWindow::mouseButtonUp(MouseButton b)
{
    handleMouseUp(b, activeTool_.get(), ... );
    return false;
}

void DrawWindow::mouseExit(MouseButton b)
{
    handleMouseUp(b, activeTool_.get(), ... );
    return false;
}
```

Reply:
done